KOGITO-6292: Improve jbpm Engine Message 1:1 (Adding correlation keys and collaboration compliant spec subscription)

Josephblt opened the pull request:
Correlation keys and collaboration are now supported by the editor.

JIRA: [KOGITO-6292](https://issues.redhat.com/browse/KOGITO-6292)

VS Code: [VSCode Extension](https://drive.google.com/file/d/1i9KzH4ZtTaqnm2uJXZeScaj4r4pV94hG/view?usp=sharing)

domhanak commented on packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/definition/property/collaboration/Correlation.java:
```diff
@@ -0,0 +1,162 @@
+    }
+
+    public static Boolean isValidString(String correlationValue) {
+        String[] correlationParts = splitCorrelationString(correlationValue);
```
I sense possible NPE here

domhanak commented:
@Josephblt Went throught the code to familiarize myself, looks good, some comment may appear, so let me know your opinion. Will test later this week to see whats the state now

romartin reviewed (commented):
Hey @Josephblt , code looks good!

Just one minor thing - does it make sense to add some help messages on the new fields for the correlation?

Thx!

domhanak commented on packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/definition/property/collaboration/Correlation.java:
```diff
@@ -0,0 +1,162 @@
+    public static Boolean isValidString(String correlationValue) {
+        String[] correlationParts = splitCorrelationString(correlationValue);
+
+        if (correlationParts.length != 5) {
```
Magic number - please rename to some constant for example `VALID_CORRELATION_PARTS_LENGTH`

domhanak commented on packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/definition/property/collaboration/Correlation.java:
```diff
@@ -0,0 +1,162 @@
+            return false;
+        }
+
+        if (!correlationParts[0].equals(IDENTIFIER)) {
```
Could we firstly parse the parts, and then check?
It will be better readable imho.
Something like:
```
correlationIdentifier = correlationParts[0];
correlationId = correlationParts[1];
correlationName = correlationParts[2]
...
```

domhanak commented on packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/definition/property/collaboration/Correlation.java:
```diff
@@ -0,0 +1,162 @@
+
+    public static Correlation fromString(String correlationValue) throws Exception {
+        if (!isValidString(correlationValue)) {
+            throw new Exception("The value: " + correlationValue + " is not a valid Correlation.");
```
Is it possible to tell user what is wrong? (Name missing or missing property..)

domhanak reviewed (commented):
some comments inline, let me know

romartin reviewed (commented):
Thanks @Josephblt , code looks good, although would be nice, as commented above, just adding the i18n mesages for showing some help message for each new field, maybe with some example. Anyway this should not block this PR, we can add a new commit on top later if necessary.

Please let me and @domhanak know once this is ready for testing. Also please guys let's take this as a priority, deadline is tomorrow (Fri). Thanks in advance!

domhanak commented:
@handreyrc @romartin I was able to configure the collaborations and correlations using the provided widgets.

There are some UX issues, where the UI shows an extra empty row:
IN following screenshot you can see the extra white row under Select value row. This also happens when I add new collaboration, the dropdown for propertyType has the extra row too.
![Screenshot from 2022-03-21 09-04-30](https://user-images.githubusercontent.com/5937251/159223680-e38e163e-2572-4484-b462-e07cccf63f45.png)

domhanak reviewed (changes requested):
I am able to add duplicate correlations. Just add correlations with same ID and the editor does not say anything to the user.

domhanak reviewed (approved):
Misunderstood the duplicate ID's, it is handled as expected. Apologies, good to go if build is green.
Please address the "ghost white row" issue if possible or let me know so that I can report it.

domhanak reviewed (changes requested):
It seems that  the correlation data is not saved (written upon save).
@Josephblt coudl you please check? I amsuing the vsix from PR checks

Josephblt commented:
Hello @domhanak, @romartin. All set. Ready to test. I created a UI validation. Invalid fields are highlighted in red. Error messages give information about what actions to take. The OK dialog button is enabled/disabled according to correlations valid/invalid states.

LuboTerifaj reviewed (changes requested):
Hello @Josephblt , 

as we discussed on chat, I reviewed the the PR and here are my findings:

###	UX issues
- The warning messages, like "_ID cannot be empty._", especially with red font is distracting. It would be better to show them only when user clicks on Add a new Correlation or tries to confirm the changes by clicking on OK.
- Having Tooltip messages next to the fields would be great.
  - I understand it is not part of the requirement, so take this just as a possible future improvement.
- UX designers input might be beneficial here.

### Minor Bugs
- The redundant empty row in Property's and Property Type's drop-down menu is still present. However, this is probably expected as it appears also in other components in Process editor.	
		
### Blocker Bugs
- It's not possible to set `Custom` type under `Property Type`.	
- No Message Correlation properties can be saved.
  - Steps to reproduce:
    1. Add Correlation in Process Properties panel.
    2. Set up all Correlation properties in Message Properties panel.
    3. Save, close and open the process.
	
### BPMN file observation
I have also checked the bpmn xml file and it doesn't conform with description in jira:	
- `bpmn2:correlationProperty id="{Property ID}" name="{Property Name}" type="{itemDefintion.id(Property Type)}"/>` in jira vs `<bpmn2:correlationProperty id="{Property ID}" name="{Property Name}" type="{Property ID}_Type">` in xml
  - I am not sure what was meant by `{itemDefintion.id(Property Type)}` expression here. The `Property Type` in parentheses is a little bit confusing. Can you please confirm the actual form of the type value (`{Property ID}_Type`) taken from xml is correct?
- `itemDefinition id="{Property ID}TypeDef"` in jira vs `<bpmn2:itemDefinition id="{Property ID}_Type"` in xml
- `messageRef` attribute is completely missing in xml
- `message` element is completely missing in xml
- `messagePath` and `dataPath` are different in xml in comparison with jira
	
### Duplicate IDs issues
The behaviour with duplicate IDs and different names is not ideal.
- The message "_Duplicate IDs with different names are not allowed._" is shown also next to text boxes that are not currently being edited. Maybe it would be better to highlight them and show the message only next to the text box that is edited.
- If you add a new Correlation with already existing ID, you can set the name which shouldn't be possible. The name could be filled automatically.
- If you add a new Correlation with already existing ID and already existing name, you are not allowed to edit the name until you update the ID first, but if you add another Correlation with already existing ID, you can edit all Names that have the same ID. It may be confusing or even irritating for some users.
- UX designers input might be beneficial here as well.

Thank you!

Josephblt commented:
Hello @LuboTerifaj, @romartin

I'll try to address all your comments/suggestions in order.

**UX issues**
I added a link below the table, so the user can toggle the error messages. The link only shows when there are errors. This way it becomes less invasive.

**Minor Bugs**
No changes here.

**Blocker Bugs**
- Fixed. Now it is possible to add custom properties again.
- There is a small problem here. You need to set a message for the correlation properties to save. But I admit it can get confusing. I can try and create a filter to make correlation fields read-only if there is no message set. But I don't think this is a blocker.

**BPMN file observation**
- Yes that is correct on XML. I could have explained better on the JIRA.
- Same as above.
- Message needs to be set to be saved on XML.
- Also needs a message to be on the XML.
- Could not find any differences

**Duplicate IDs issues**
- That is by design. I cannot know which of the names is the wrong one so I signal both.
- I agree, this is now implemented. So whenever you input an existing ID and the name is empty, it will set the name automatically.
- That is also by design. I think the auto name setting helps to minimize this problem. But once again you cannot determine which of the names is the right one. So all the correlations sharing an ID get signaled.

Let me know what you guys think. Thanks.

LuboTerifaj reviewed (changes requested):
Hello @Josephblt ,

Thank you for the fixes. Here are my comments:

### UX issues
- Show/Hide errors is a nice functionality, but its behaviour is not ideal. If there is an issue during adding correlation, the message should be always shown to the user without additional need to click on Show/Hide Errors button.
- As I already mentioned, UX designer input would be beneficial here.

### Blocker bugs
- "No Message Correlation properties can be saved." issue is not fixed. See the steps to reproduce in my previous [review](https://github.com/kiegroup/kie-tools/pull/856#pullrequestreview-920921369).

### BPMN file observation
- You clarified the first two points from my previous [review](https://github.com/kiegroup/kie-tools/pull/856#pullrequestreview-920921369) are correct.
- However the rest, `messageRef`, `message`, `messagePath` and `dataPath` issues still apply.

### Duplicate IDs issues
Thank you for the fixes. It works good now.

Josephblt commented:
Hello @LuboTerifaj, @romartin. Here is a video with all the steps. At 0:37 notice the field Message in the Implementation/Execution of the properties tab from the Start Message Node. You have to set this field to something. Otherwise the correlation properties cant link to anything. They are serialized to the BPMN file, but when you try to open they won't load.

https://user-images.githubusercontent.com/25369234/160188387-934d909e-c2ca-4e22-b9c1-6b3c4c50bf91.mp4

romartin reviewed (approved):
Thanks @Josephblt for the video , it clarifies.

Approving - The feature works, although is not so intuitive as we may expect, as some field have to be filled in some "order". Anyway, IMO we can properly document the usecase for this, also creating some quick blog post about it, and so DO NOT block this feature for the release. It results quite important for vscode and "kogito" runtimes integration.

@LuboTerifaj @domhanak missing your reviews here pleaseee, we still have some chance for merging and making this available for next release,

Thanks everyone involved guys!

LuboTerifaj commented:
Hello @Josephblt , 

Why it is necessary to set Message property in order to save all Correlation values in Message events?

I think the Correlation values should be saved even without the Message property is set, or the Correlation property should be disabled completely to avoid losing the filled data.

Thanks!

Josephblt commented:
Hello @LuboTerifaj. We need to look at the BPMN file to understand this. Let's consider this message data example.
```
<bpmn2:message id="_5i1JYJGKEDqHtJYbEuZ8UA" itemRef="MessageType" name="Message"/>

<bpmn2:correlationProperty id="Property_ID_1" name="Property Name 1" type="Property_ID_1_Type">
     <bpmn2:correlationPropertyRetrievalExpression id="_7CE42BE1-45FB-4189-96B8-89649C395487" messageRef="_5i0iUJGKEDqHtJYbEuZ8UA">
          <bpmn2:messagePath id="_AD69640F-DCEA-46BD-9BC0-17406A47CB73" evaluatesToTypeRef="_741E4D8C-DA2E-4044-8EF9-707685E50DEC" language="java">MessageExpression</bpmn2:messagePath>
     </bpmn2:correlationPropertyRetrievalExpression>
</bpmn2:correlationProperty>
```
In the correlationPropertyRetrievalExpression node what binds everything together is the messageRef attribute. This messageRef has nothing to do with the Start/Intermediate Message Event. It is simply used by it. So both the event and the correlationProperty share a reference to it. Hence why a Message is needed to link them together.

Now that I look, I agree that the UI arrangement is not ideal. I'm open to suggestions.

LuboTerifaj reviewed (approved):
Hello @Josephblt ,

I have reported the issue with error messages here: https://issues.redhat.com/browse/KOGITO-6920

As we discussed on chat, let's address the critical bug related to not saved Correlation properties in the follow-up jira: https://issues.redhat.com/browse/KOGITO-6919. However, I recommend to fix it ASAP.

Regarding BPMN file, `messagePath` and `dataPath` lacks `xsi:type="bpmn2:tFormalExpression"`, but I am not sure if this is an issue.

Thank you!
